### ISSUE TITLE
Go To Definition.

### ISSUE THREAD

**Issue by mfcd** [JunoLab/atom-julia-client#76]
Hi,

New to Juno and new to Atom.
Trying to get some sort of Go To Definition.

I have tried to configure the `goto` package but not sure how to proceed.
Due to performance issues, I'd like to avoid `Hydrogen`

Thanks.

**Comment by spencerlyon2** [comment]
I use the [`atom-ctags`](https://atom.io/packages/atom-ctags) and have [this content](https://github.com/JuliaLang/julia/blob/master/contrib/ctags) my `~/.ctags` file.

There are some rough edges, but it does work pretty well

**Comment by Varanas** [comment]
You can also use the `Toggle methods` command (at least on master), which will display all the method for any given function call with clickable links to their respective definitions.

We should probably use an interface similar to that of `atom-ctags` though, I like it -- doesn't require a mouse.

**Comment by spencerlyon2** [comment]
Good point!

Could we have a command that gathers the info from `Toggle methods`, puts it in a modal (like you get from `cmd+shift+p` on osx) and lets the user select which method they want to visit?

**Comment by Varanas** [comment]
Yeah, shouldn't be too much trouble to implement. I'll have a look.

**Comment by Varanas** [comment]
Oh well, not as easy as I thought. I've made some progress, but haven't quite got this working.

In principle though this is sooo much nicer than my old implementation -- thanks for bringing it up!

**Comment by Varanas** [comment]
Success:
![methodsview](https://cloud.githubusercontent.com/assets/6735977/12198592/c254b48a-b611-11e5-857a-5371eac149df.png)
Needs a bit of cleaning up and nicer display (something like the path or at least the file any given method's defined in), but basically works.

**Comment by spencerlyon2** [comment]
Nice, that looks like a great start. 

I would also vote that you have the path

**Comment by spencerlyon2** [comment]
This is the UX from `atom-ctags`.

<img width="452" alt="screen shot 2016-01-08 at 8 46 30 am" src="https://cloud.githubusercontent.com/assets/1001948/12201948/53791a56-b5e4-11e5-80e4-d7f23b0e7793.png">


I think that they do a good job with the line number and file - but I like the way you handled types. Maybe we could implement something similar for method location on top of what you have? Also can we drop the margins on the modal?

**Comment by Varanas** [comment]
Sure, that's just style stuff and shouldn't be too hard.

I'm just a bit conflicted on how to handle the intercommunication from Julia to Atom for things like this where we actually return data instead of DOM nodes. Either I implement it how I think it should be done (construct an array of `method`-objects in Julia that contains method signature, filename and line number), duplicate quite a lot of effort in `Atom.jl´ and do it completly different than what `Atom.jl` does otherwise, or I try to hack something together to extract the path from the dom node or something.

Basically I'm not sure if the current apporach of doing most of the rendering in `Atom.jl` is a good approach -- it's pretty different from what I'd understand as the usual MVC-approach.

`/rambling-over`

**Comment by spencerlyon2** [comment]
I'm a little confused. 

I would imagine something like this: 

- atom-julia-client asks Atom.jl for method info (e.g. calls `methods(func_name)` for the `func_name` under cursor)
- Atom.jl then gives atom-julia-client a vector of 3-tuples: (method signature, file, line)
- atom-julia-client then throws them into a stylized modal.

This might be what you are already doing, I'm not sure. Does that not seem to work for some reason?

**Comment by Varanas** [comment]
Yeah, that's exactly what I'm doing. It's just different from what the rest of Juno is doing, where we do pretty much all the rendering in Julia and pass the complete dom nodes to Atom. And because of that, most methods deal with those nodes -- setting up file links is annoying because of that for example.

**Comment by kindlychung** [comment]
On a modern IDE, people usually expect to ctrl-click on a function/variable to jump to its definition.
